File: typescript/day02.ts
import { readFileSync } from 'fs';
import { match, P } from 'ts-pattern';

const inputFilename: string = '../data/input02.txt';
const lines = readFileSync(inputFilename, 'utf8').split('\n');

type Shape = 1 | 2 | 3;
type StrategyKey = 'A' | 'B' | 'C' | 'X' | 'Y' | 'Z';

const Rock: Shape = 1;
const Paper: Shape = 2;
const Scissors: Shape = 3;

const Strategy: Record<StrategyKey | string, Shape> = {
  A: Rock,
  B: Paper,
  C: Scissors,
  X: Rock,
  Y: Paper,
  Z: Scissors,
};

type Left = 0;
type Right = 1;
type Draw = -1;

type Winner = Left | Right | Draw;
type Round = [Shape, Shape];

const getWinner = (round: Round): Winner =>
  match(round)
    // Draw
    .with([Rock, Rock], () => -1 as Draw)
    .with([Paper, Paper], () => -1 as Draw)
    .with([Scissors, Scissors], () => -1 as Draw)
    // Left wins
    .with([Rock, Scissors], () => 0 as Left)
    .with([Paper, Rock], () => 0 as Left)
    .with([Scissors, Paper], () => 0 as Left)
    // Right wins
    .with([Rock, Paper], () => 1 as Right)
    .with([Paper, Scissors], () => 1 as Right)
    .with([Scissors, Rock], () => 1 as Right)
    .exhaustive();

const getStrategyShape = (key: string): Shape | undefined => Strategy[key];

const rounds: Round[] = lines.reduce((acc, line) => {
  const [x, y] = line.split(' ');
  const [a, b] = [getStrategyShape(x), getStrategyShape(y)];

  if (a && b) {
    acc.push([a, b]);
  }

  return acc;
}, [] as Round[]);

// The score for a single round is
//
//   1. The value for the shape you selected
//   2. Plus the score for the outcome of the round
//     - 0 if you lost
//     - 3 if the round was a draw
//     - 6 if you won
//
const calcPoints = (rounds: Round[]) =>
  rounds.map(([left, right]) => {
    const winner = getWinner([left, right]);

    const points =
      right +
      match(winner)
        .with(0 as Left, () => 0)
        .with(1 as Right, () => 6)
        .with(-1 as Draw, () => 3)
        .exhaustive();

    return points;
  });

// The score for a single round is
//
//   1. The value for the shape you selected
//   2. Plus the score for the outcome of the round
//     - 0 if you lost
//     - 3 if the round was a draw
//     - 6 if you won
//
const points = calcPoints(rounds);
const totalScore = points.reduce((sum, n) => sum + n, 0);

console.log({
  'rounds[0..5]': rounds.slice(0, 5),
  'points[0...5]': points.slice(0, 5),
  totalScore,
});

// Part Two

// Get the Shape to produce given Winner against given Shape
const getShapeFor = (left: Shape, winner: Winner): Shape =>
  match([left, winner] as [Shape, Winner])
    // Draw
    .with([Rock, -1 as Draw], () => Rock)
    .with([Paper, -1 as Draw], () => Paper)
    .with([Scissors, -1 as Draw], () => Scissors)
    // Left wins
    .with([Rock, 0 as Left], () => Scissors)
    .with([Paper, 0 as Left], () => Rock)
    .with([Scissors, 0 as Left], () => Paper)
    // Right wins
    .with([Rock, 1 as Right], () => Paper)
    .with([Paper, 1 as Right], () => Scissors)
    .with([Scissors, 1 as Right], () => Rock)
    .exhaustive();

// the second column says how the round needs to end:
//   - X means you need to lose
//   - Y means you need to end the round in a draw
//   - Z means you need to win
const rounds2: Round[] = lines.reduce((acc, line) => {
  const [x, y] = line.split(' ');

  const left = getStrategyShape(x);

  if (left) {
    const desiredWinner: Winner = match(y)
      .with('X', () => 0 as Left)
      .with('Y', () => -1 as Draw)
      .with('Z', () => 1 as Right)
      .run();

    const right = getShapeFor(left, desiredWinner);
    acc.push([left, right]);
  }

  return acc;
}, [] as Round[]);

const points2 = calcPoints(rounds2);
const totalScore2 = points2.reduce((sum, n) => sum + n, 0);

console.log({
  'rounds2[0..5]': rounds2.slice(0, 5),
  'points2[0...5]': points2.slice(0, 5),
  totalScore2,
});

File: rust/src/bin/day02.rs
use std::fs;

#[derive(Debug, Clone)]
enum Shape {
    Rock = 1,
    Paper = 2,
    Scissors = 3,
}

#[derive(Debug, Clone)]
enum Winner {
    Left = 0,
    Right = 6,
    Draw = 3,
}

type Code = String;
type Pair = (Code, Code);
type Round = (Shape, Shape);

fn get_left_shape(code: Code) -> Option<Shape> {
    return match code.as_str() {
        "A" => Some(Shape::Rock),
        "B" => Some(Shape::Paper),
        "C" => Some(Shape::Scissors),
        _ => None,
    };
}

fn get_right_shape(code: Code) -> Option<Shape> {
    return match code.as_str() {
        "X" => Some(Shape::Rock),
        "Y" => Some(Shape::Paper),
        "Z" => Some(Shape::Scissors),
        _ => None,
    };
}

// get (shape, shape) for given pair of codes
fn get_round((l, r): Pair) -> Option<Round> {
    return match (get_left_shape(l), get_right_shape(r)) {
        (Some(left), Some(right)) => Some((left, right)),
        _ => None,
    };
}

// get round to produce given winner against given left shape
fn get_round_for(left: Shape, winner: Winner) -> Round {
    let right = match (left.clone(), winner) {
        // Left
        (Shape::Rock, Winner::Left) => Shape::Scissors,
        (Shape::Paper, Winner::Left) => Shape::Rock,
        (Shape::Scissors, Winner::Left) => Shape::Paper,
        // Right
        (Shape::Rock, Winner::Right) => Shape::Paper,
        (Shape::Paper, Winner::Right) => Shape::Scissors,
        (Shape::Scissors, Winner::Right) => Shape::Rock,
        // Draw
        (Shape::Rock, Winner::Draw) => Shape::Rock,
        (Shape::Paper, Winner::Draw) => Shape::Paper,
        (Shape::Scissors, Winner::Draw) => Shape::Scissors,
    };

    return (left, right);
}

fn get_winner(round: Round) -> Winner {
    match round {
        // Left
        (Shape::Rock, Shape::Scissors) => Winner::Left,
        (Shape::Paper, Shape::Rock) => Winner::Left,
        (Shape::Scissors, Shape::Paper) => Winner::Left,
        // Right
        (Shape::Rock, Shape::Paper) => Winner::Right,
        (Shape::Paper, Shape::Scissors) => Winner::Right,
        (Shape::Scissors, Shape::Rock) => Winner::Right,
        // Draw
        (Shape::Rock, Shape::Rock) => Winner::Draw,
        (Shape::Paper, Shape::Paper) => Winner::Draw,
        (Shape::Scissors, Shape::Scissors) => Winner::Draw,
    }
}

// The score for a single round is
//
//   1. The value for the shape you selected
//   2. Plus the score for the outcome of the round
//     - 0 if you lost
//     - 3 if the round was a draw
//     - 6 if you won
//
fn get_points((left, right): Round) -> u32 {
    let shape_points = right as u32;
    let round_points = get_winner((left, right)) as u32;
    return shape_points + round_points;
}

fn main() {
    let input_filename = "../data/input02.txt";

    // vec of lines from file
    let lines: Vec<String> = fs
        ::read_to_string(input_filename)
        .expect("Failed to read input")
        .trim()
        .split("\n")
        .map(String::from)
        .collect();

    // vec of (string, string) pairs of codes
    let mut pairs: Vec<Pair> = Vec::new();
    for line in lines.iter() {
        let mut moves: Vec<String> = line.to_string().split(" ").map(String::from).collect();

        if moves.len() >= 2 {
            let left = moves.remove(0);
            let right = moves.remove(0);
            pairs.push((left, right));
        }
    }

    // vec of (shape, shape) pairs
    let mut rounds: Vec<Round> = Vec::new();
    for (l, r) in pairs.iter() {
        if let Some(round) = get_round((l.to_string(), r.to_string())) {
            rounds.push(round);
        }
    }

    // calculate points
    let mut points: Vec<u32> = Vec::new();
    for (left, right) in rounds {
        let round_points = get_points((left, right));
        points.push(round_points);
    }

    let total: u32 = points.iter().sum();

    // Round Two

    // the second column says how the round needs to end:
    //   - X means you need to lose
    //   - Y means you need to end the round in a draw
    //   - Z means you need to win
    let mut rounds2: Vec<Round> = Vec::new();
    for (l, r) in pairs.iter() {
        let left = get_left_shape(l.to_string()).unwrap();

        #[allow(unused_parens)]
        if
            let Some(desired_winner) = (match r.as_str() {
                "X" => Some(Winner::Left),
                "Y" => Some(Winner::Draw),
                "Z" => Some(Winner::Right),
                _ => None,
            })
        {
            let round = get_round_for(left, desired_winner);
            rounds2.push(round);
        }
    }

    // calculate points
    let mut points2: Vec<u32> = Vec::new();
    for (left, right) in rounds2 {
        let round_points = get_points((left, right));
        points2.push(round_points);
    }

    let total2: u32 = points2.iter().sum();

    println!("{:?}, {:?}", total, total2);
}
